DDUF parser v0.1

Wauplin opened the pull request:
Related to https://github.com/huggingface/huggingface.js/tree/main/packages/dduf/src and https://github.com/huggingface/diffusers/pull/10037.

Goal is to centralize the common bricks to save and load a DDUF file, therefore ensuring the specs are followed (e.g. force ZIP64, no compression, only txt/json/gguf/safetensors files). For now the implementation is based on the built-in `zipfile` module for practical reasons.

**TODO:**
- [x] high priority: do not rely on private attribute to determine file offset in zipfile (see https://github.com/huggingface/diffusers/pull/10037#discussion_r1874795428)
- [x] ~update `write_dduf_file` signature to infer the DDUF filename based on model name, quantization, etc. See https://github.com/huggingface/diffusers/pull/10037#discussion_r1862275730~ 
=> let's not do it in the end https://github.com/huggingface/diffusers/pull/10037#discussion_r1869315926
- [ ] optimize when writing file (at the moment, read files chunk by chunk and add them to the zip => there might be a quicker way to do that)
- [x] allow writing directly from blobs => would avoid having to save as safetensors on disk +  re-save as DDUF
- [x] add proper tests

**Short docs:** https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_2692/en/package_reference/serialization#dduf-file-format

**Example:** https://huggingface.co/Wauplin/stable-diffusion-v1-4-DDUF

### How to write a DDUF file?

```python
>>> from huggingface_hub import write_dduf_file
>>> export_folder_as_dduf("FLUX.1-dev.dduf", folder_path="path/to/FLUX.1-dev")
```

### How to read a DDUF file?

```python
>>> import json
>>> import safetensors.load
>>> from huggingface_hub import read_dduf_file

# Read DDUF metadata
>>> dduf_entries = read_dduf_file("FLUX.1-dev.dduf")

# Returns a mapping filename <> DDUFEntry
>>> dduf_entries["model_index.json"]
DDUFEntry(filename='model_index.json', offset=66, length=587)

# Load model index as JSON
>>> json.loads(dduf_entries["model_index.json"].read_text())
{'_class_name': 'FluxPipeline', '_diffusers_version': '0.32.0.dev0', '_name_or_path': 'black-forest-labs/FLUX.1-dev', 'scheduler': ['diffusers', 'FlowMatchEulerDiscreteScheduler'], 'text_encoder': ['transformers', 'CLIPTextModel'], 'text_encoder_2': ['transformers', 'T5EncoderModel'], 'tokenizer': ['transformers', 'CLIPTokenizer'], 'tokenizer_2': ['transformers', 'T5TokenizerFast'], 'transformer': ['diffusers', 'FluxTransformer2DModel'], 'vae': ['diffusers', 'AutoencoderKL']}

# Load VAE weights using safetensors
>>> with dduf_entries["vae/diffusion_pytorch_model.safetensors"].as_mmap() as mm:
...     state_dict = safetensors.torch.load(mm)
```

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_2692). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

julien-c reviewed (approved):
clean v0 implementation!

Should work well enough to prototype+validate the format with partners. cc @SunMarc @LysandreJik

Wauplin commented on src/huggingface_hub/serialization/_dduf.py:
```diff
@@ -0,0 +1,142 @@
+    return entries
+
+
+def write_dduf_file(dduf_path: Union[str, Path], diffuser_path: Union[str, Path]) -> None:
```
Or even `entries` could be a `Iterable[Tuple[str, Union[str, Path, bytes, BinaryIO]]`. This way you could pass an iterator that will serialize things on the fly (no need to have everything in memory at once).

sayakpaul replied:
```
write_dduf_file("FLUX.1-dev.dduf", diffuser_path="path/to/FLUX.1-dev")
```
here could the `diffuser_path` be renamed into something more general as we want DDUF to be independent of `diffusers`.

SunMarc replied:
It would be nice to see how we can blend this with the current logic of diffusers : https://github.com/huggingface/diffusers/pull/10037/files#:~:text=**save_kwargs)-,if%20dduf_filename%3A,-SunMarc%20marked%20this

Right now, I'm saving the files to the archive right after each model gets saved

Wauplin replied:
Done in https://github.com/huggingface/huggingface_hub/pull/2692/commits/1b11f0b38c71c3d6ad3e1a7adfb9d2956e219741

Now you can pass a list of files to export:

```python
# Export specific files from the local disk.
>>> from huggingface_hub import export_entries_as_dduf
>>> export_entries_as_dduf(
...     "stable-diffusion-v1-4-FP16.dduf",
...     entries=[ # List entries to add to the DDUF file (here, only FP16 weights)
...         ("model_index.json", "path/to/model_index.json"),
...         ("vae/config.json", "path/to/vae/config.json"),
...         ("vae/diffusion_pytorch_model.fp16.safetensors", "path/to/vae/diffusion_pytorch_model.fp16.safetensors"),
...         ("text_encoder/config.json", "path/to/text_encoder/config.json"),
...         ("text_encoder/model.fp16.safetensors", "path/to/text_encoder/model.fp16.safetensors"),
...         # ... add more entries here
...     ]
... )
```

or even iterate over entries to save to lazily dump state dicts without saving to disk:

```python
# Export state_dicts one by one from a loaded pipeline 
>>> from diffusers import DiffusionPipeline
>>> from typing import Generator, Tuple
>>> import safetensors.torch
>>> from huggingface_hub import export_entries_as_dduf
>>> pipe = DiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4")
... # ... do some work with the pipeline

>>> def as_entries(pipe: DiffusionPipeline) -> Generator[Tuple[str, bytes], None, None]:
...     # Build an generator that yields the entries to add to the DDUF file.
...     # The first element of the tuple is the filename in the DDUF archive (must use UNIX separator!). The second element is the content of the file.
...     # Entries will be evaluated lazily when the DDUF file is created (only 1 entry is loaded in memory at a time)
...     yield "vae/config.json", pipe.vae.to_json_string().encode()
...     yield "vae/diffusion_pytorch_model.safetensors", safetensors.torch.save(pipe.vae.state_dict())
...     yield "text_encoder/config.json", pipe.text_encoder.config.to_json_string().encode()
...     yield "text_encoder/model.safetensors", safetensors.torch.save(pipe.text_encoder.state_dict())
...     # ... add more entries here

>>> export_entries_as_dduf("stable-diffusion-v1-4.dduf", as_entries=as_entries(pipe))
```

And I've kept exporting an entire folder at once for ease of use:

```python
# Export a folder as a DDUF file
>>> from huggingface_hub import export_folder_as_dduf
>>> export_folder_as_dduf("FLUX.1-dev.dduf", folder_path="path/to/FLUX.1-dev")
```

Wauplin commented on src/huggingface_hub/serialization/_dduf.py:
```diff
@@ -0,0 +1,142 @@
+    #       Example: "stable-diffusion-3.5-Q4-BNB.dduf"
+    #       See https://github.com/huggingface/diffusers/pull/10037#discussion_r1862275730.
+    logger.info("Writing DDUF file %s from folder %s", dduf_path, diffuser_path)
+    diffuser_path = Path(diffuser_path)
```
yes, I'm thinking about updating the API: https://github.com/huggingface/huggingface_hub/pull/2692#discussion_r1869330886

Wauplin replied:
done (see docs). You can now pass a list of entries.

sayakpaul commented on src/huggingface_hub/serialization/_dduf.py:
```diff
@@ -0,0 +1,142 @@
+
+logger = logging.getLogger(__name__)
+
+DDUF_ALLOWED_ENTRIES = {".json", ".gguf", ".txt", ".safetensors"}
```
Let's add ".model", ".bin", ".pt" as well? 

Most of the ecosystem libraries now enforce a `weights_only=True` when using `torch.load()` -- so the PyTorch pickle should probably be okay to add here. 

`.model` is for the sentence-piece tokenizer: 
https://huggingface.co/black-forest-labs/FLUX.1-dev/blob/main/tokenizer_2/spiece.model

julien-c replied:
> ".bin", ".pt" as well?

i think we said we deliberately don't want to support legacy pickle otherwise the format isn't secure

julien-c replied:
i think it's a good constraint to put

sayakpaul replied:
Okay. .model is still needed.

Wauplin replied:
Done adding  `.model` to the list

sayakpaul commented on src/huggingface_hub/serialization/_dduf.py:
```diff
@@ -0,0 +1,142 @@
+
+    Only the metadata is read, the data is not loaded in memory.
+
+    Args:
```
(nit): an example here as well?

sayakpaul commented on src/huggingface_hub/serialization/_dduf.py:
```diff
@@ -0,0 +1,142 @@
+    diffuser_path = Path(diffuser_path)
+    with zipfile.ZipFile(str(dduf_path), "w", zipfile.ZIP_STORED) as archive:
+        for path in diffuser_path.glob("**/*"):
+            if path.is_dir():
+                logger.debug("Skipping directory %s", path)
+                continue
```
Why are we skipping directories? A `diffusers` formatted pipeline can have directories:
https://huggingface.co/black-forest-labs/FLUX.1-dev/tree/main

Or am I misunderstanding this?

sayakpaul replied:
Ah thanks for explaining

sayakpaul reviewed (commented):
Thanks for the neat examples and starting this!

SunMarc commented on src/huggingface_hub/serialization/_dduf.py:
```diff
@@ -0,0 +1,142 @@
+    entries = {}
+    dduf_path = Path(dduf_path)
+    logger.info("Reading DDUF file %s", dduf_path)
+    with zipfile.ZipFile(str(dduf_path), "r") as zf:
+        for info in zf.infolist():
```
This can be used to check if the files are corrupted : `zf.testzip()`

Wauplin replied:
For now I'll skip validation (see https://github.com/huggingface/huggingface_hub/pull/2692#discussion_r1869931183) but agree for a v1

Wauplin commented on src/huggingface_hub/serialization/_dduf.py:
```diff
@@ -0,0 +1,142 @@
+    with zipfile.ZipFile(str(dduf_path), "r") as zf:
+        for info in zf.infolist():
+            logger.debug("Reading entry %s", info.filename)
+            if info.compress_type != zipfile.ZIP_STORED:
```
for now I'll let the backend enforce the data validation and will propagate in huggingface_hub once settled (I prefer to add all the validation at once).
Here I'm checking the compress_type because otherwise it would lead to an immediate error when reading the file (contrary to zip64 check that is not necessary to read data).

hanouticelina replied:
I'm maybe lacking a bit of context, but i'm guessing the choice of 0 compression is to ensure that the deduplication works right? are there any other reasons ?

Wauplin replied:
Yes, when loading a safetensors file from the dduf file, we want to be able to do a memory-mapping of the chunk corresponding to the safetensors data and then pass that to `safetensors.torch.load`. Otherwise we would have to decompress, load the safetensors file in memory and then load from there.

Wauplin commented on src/huggingface_hub/serialization/_dduf.py:
```diff
@@ -0,0 +1,142 @@
+    # TODO: update method signature.
+    #       DDUF filename should be inferred as much as possible from high-level info (precision, model, etc.) to ensure consistency.
+    #       Example: "stable-diffusion-3.5-Q4-BNB.dduf"
+    #       See https://github.com/huggingface/diffusers/pull/10037#discussion_r1862275730.
```
removed ti

rwightman commented:
neat packaging idea... not sure what the highest priority is but a consideration re zip vs tar, tar would let you start reading weights into memory w/o seeking to end and having to access to the zip directory (in footer). To mitigate the direct access to offsets (w/o reading forward) in tar, you can write a TOC metadata file at the start of the tar to work around that. Like uncompressed zip you have direct access to bytes in a tar file.

julien-c commented:
cc @ylow maybe for context on why we opted to go with zip vs. tar^

ylow commented:
@rwightman zip or tar, either way an index structure will be necessary. And zip comes with one as part of the format so no need to invent another. Indeed it might seem that one additional "seek" might be necessary to obtain the index structure, but say from a HTTP range read you can do "Range: bytes=-100000" for instance to fetch the last 100k, so in that sense there is no major difference reading from start or end from a streaming perspective. On the other hand its kinda nice to be able to rely on a common container format so that minimal effort is needed to implement readers in other languages.

rwightman commented:
@ylow FWIW that extra range request is a seek and an extra network RTT (or few), one of the reasons why reading from parquet files when 'streaming' using fsspec is quite a bit slower at scale than just streaming tar (forward streaming only). The server does pay a penalty with extra IOPS. 

Probably not as big an issue with model weights vs dataset files...

ylow commented:
@rwightman Would love to try to understand some of the details here since it sounds like you have encountered similar issues. I am thinking if the index is in a header, I will need to issue 1 Range query for say... the first 64KB to fetch it. And if the index is in a footer, I similarly will only need to issue 1 Range query for the last 64KB to fetch it. (Yes technically you read the last few bytes first to get the length of the footer, then read the footer. But I can always optimistically fetch a bigger block). So there shouldn't be any additional RTT involved?

rwightman commented:
@ylow you don't need range requests if the underlying file is 'streamable', you can just start transfering and parse as the bytes are received. It involves the fewest network RTT.

On the storage side, moving through the file from start to end in sequence is by far the most efficient with many storage arch because it's ameanable to typical prefetch caching mechanisms without jumping around the file. Doing range requests in sequence won't be much different than doing one transfer, but going to the end of the file and back (or jumping around across many blocks randomly) will add overhead unless you know you are always on fast SSD backed storage and don't need to rely on prefetch/buffering to smooth latency.

sayakpaul commented on src/huggingface_hub/serialization/_dduf.py:
```diff
@@ -0,0 +1,289 @@
+
+logger = logging.getLogger(__name__)
+
+DDUF_ALLOWED_ENTRIES = {
+    ".gguf",
+    ".json",
+    ".model",
+    ".safetensors",
+    ".txt",
+}
```
I am okay with this setup but just also saying as an FYI that `torchao` currently doesn't support `safetensors` serialization. So, any `diffusers` repo serialized with `torchao` won't make the cut here. But we can probably worry about that later.

Wauplin replied:
Might be a good incentive for them to support safetensors :)

sayakpaul commented on src/huggingface_hub/serialization/_dduf.py:
```diff
@@ -0,0 +1,289 @@
+
+def add_entry_to_dduf(dduf_path: Union[str, Path], filename: str, content: Union[str, Path, bytes]) -> None:
+    """
+    Add an entry to an existing DDUF file.
```
(nit): maybe a note on when this is useful and an example?

Wauplin replied:
actually not sure it is useful in the end :smile: I used it during my testing and I do see a possibility where we would like to take an existing DDUF file and add an adapter file to it for instance. That's why I kept it.

I did not document it nor exposed it in `huggingface_hub`

Wauplin commented:
@rwightman @ylow Re: tar vs zip and range requests vs stream: those are interesting topics and depending on the context, we could prefer one or the other. In our case here, we are interested in:
- been able to read the index from remote
=> will allow us to display metadata/index on the Hub, as done for safetensors/gguf file at the moment
=> Zip provides a TOC by default, tar could provide a TOC as well (format to be defined)
- been able to read the index from local disk + load files with memory-mapping
=> both can do it since we don't compress data
- been able to write the gguf archive file by file without knowing in advance the number of files nor their file sizes
=> this is useful if we want to export a diffusers model "on the fly" and not from a local folder on disk.
=> Zip allows us to do that very easily as the TOC is at the end of the file. For Tar files, we would need to define the header format in such a way that we can add files to it up to a certain limit. Doable but potentially more error-prone and not out-of-the-box.

However we are less interested in downloading only parts of a DDUF file, or in streaming the model from remote. Even if we do so, the number of range requests would still be quite limited (worse case, `2*N+1`, 1 for TOC + 1 per file header + 1 per file). The problem with `fsspec` is when it downloads by chunks of 5MB (I think ?) by default, hence the big overhead. As you said, this is more a problem when streaming a dataset rather than when loading a model.

Hopes this clarifies the choice!

julien-c commented:
@rwightman @ylow and BTW we already have a v1 for a remote-parser (over HTTP ranges) of DDUF, here: https://github.com/huggingface/huggingface.js/pull/1058 (similar to what we have for Safetensors and GGUF)

SunMarc commented on src/huggingface_hub/serialization/_dduf.py:
````diff
@@ -0,0 +1,289 @@
+            with mmap.mmap(f.fileno(), length=0, access=mmap.ACCESS_READ) as mm:
+                yield mm[self.offset : self.offset + self.length]
+
+    def read_text(self, encoding="utf-8") -> str:
+        """Read the file as text.
+
+        Useful for '.txt' and '.json' entries.
+
+        Example:
+            ```py
+            >>> import json
+            >>> index = json.loads(entry.read_text())
+            ```
+        """
+        with self.dduf_path.open("rb") as f:
+            f.seek(self.offset)
+            return f.read(self.length).decode(encoding=encoding)
````
Can you add a `read` method also without the decode part ? I need that to make it work on transformers side.

Wauplin replied:
btw, I'm working on new utilities to make the integrations better (see [slack](https://huggingface.slack.com/archives/C0821RXC9T5/p1733412799640689?thread_ts=1733372897.316429&cid=C0821RXC9T5) -private link-). Will keep you posted as it will definitely impact that `transformers`'s PR

SunMarc replied:
sounds good !

SunMarc commented on docs/source/en/package_reference/serialization.md:
```diff
@@ -6,6 +6,114 @@ rendered properly in your Markdown viewer.
+...     yield "text_encoder/model.safetensors", safetensors.torch.save(pipe.text_encoder.state_dict())
+...     # ... add more entries here
+
+>>> export_entries_as_dduf("stable-diffusion-v1-4.dduf", as_entries=as_entries(pipe))
```
```suggestion
>>> export_entries_as_dduf("stable-diffusion-v1-4.dduf", entries=as_entries(pipe))
```

SunMarc commented on src/huggingface_hub/serialization/_dduf.py:
```diff
@@ -0,0 +1,289 @@
+        ...     yield "text_encoder/model.safetensors", safetensors.torch.save(pipe.text_encoder.state_dict())
+        ...     # ... add more entries here
+
+        >>> export_entries_as_dduf("stable-diffusion-v1-4.dduf", as_entries=as_entries(pipe))
```
```suggestion
        >>> export_entries_as_dduf("stable-diffusion-v1-4.dduf", entries=as_entries(pipe))
```

SunMarc commented:
For serialization in diffusers, even if the utils `export_entries_as_dduf`, it is not very easy be able to serialize on the fly as we don't necessarily knows what to serialize. Diffusers call the serialization function from different library (mostly diffusers and transformers) directly so I don't know which object to serialize. 
`save_method(os.path.join(save_directory, pipeline_component_name), **save_kwargs)`
Also each object (config, generation) have their own saving logic. 

For now, I just replace the current logic (save normally then export to dduf for each model) I have in place with your api. I had to do some modification so if you can add them to the current PR, that would be nice. 
https://github.com/huggingface/huggingface_hub/compare/dduf-parser-v0.1...dduf-parser-more-feature?expand=1.
Last thing that I didn't had the time to look into is duplicates, I know that if a filename already exist in the achieve, it is not going to overwrite but hide it instead. Do you know what can be done without having to recreate the entire archive ?

Wauplin commented:
> For serialization in diffusers, even with the utils export_entries_as_dduf, it is not very easy be able to serialize on the fly as we don't necessarily know what to serialize. Diffusers call the serialization function from different library (mostly diffusers and transformers) directly so I don't know which object to serialize.

Thanks for the changes and looking into it @SunMarc! The more I think about it, the more I believe that for a v0 we should focus on:
- having helpers in `huggingface_hub` to load / save DDUF files
- making `diffusers` able to seamlessly load from a DDUF file
- post-pone `diffusers`'s optimized serialization

My reasoning behind this is that 1. we want to ship a v0 as quick as possible to gather feedback on the DDUF format 2. we are still unsure how to export to DDUF efficiently from diffusers (cf [your comment](https://github.com/huggingface/huggingface_hub/pull/2692#issuecomment-2528405146)) and 3. we don't really _need_ an optimized export method for the launch.

In my opinion what we should do is:
1. create a fork branch from https://github.com/huggingface/diffusers/pull/10037/ that only keeps the loading logic and ship it
2. ship `huggingface_hub`'s part as it is done now
3. document the current format and how to use it (load/save). Documenting `Pipeline.save_pretrained` + `export_folder_as_dduf` is enough for now.
4. release, make comms about it, get feedback about the DDUF format, etc.
5. (once the above is done), get back to the saving strategy including your changes from [dduf-parser-more-feature](https://github.com/huggingface/huggingface_hub/compare/dduf-parser-v0.1...dduf-parser-more-feature?expand=1). As it is less of a hurry, I'd rather wait a bit before adding new parameters to the main DDUF helpers.

Let me know what you think!

e(cc @sayakpaul @LysandreJik as well)

sayakpaul commented:
Agree with the overall strategy of not shipping an all around optimized v0, making sure of explicit documentation. 

> create a fork branch from https://github.com/huggingface/diffusers/pull/10037 that only keeps the loading logic and ship it

Just so that I understand we’re not agreeing to just ship _loading_ for the v0 right (it’s both saving and loading)? If so, I don’t understand the comment fully.

SunMarc commented:
Sounds good to me for this plan !

sayakpaul commented:
Then I am not sure how we could get concrete feedback with just DDUF loading for v0. 

I think users would want to experiment with both saving and loading, especially folks who want to integrate DDUF in external applications like LLM Studio. 

But maybe I am missing something so, happy to be rectified.

julien-c commented:
I agree we should focus on just loading

DDUF saving can be done as a simple script or command-line on top of existing diffusers files (ie. just packaging an existing diffusers repo)

Wauplin commented:
Yes sorry I thing we are on the same page but misunderstanding each other. What I suggest is:
- remove the dduf-saving logic from diffusers
- keep the dduf-loading logic in diffusers
- keep the dduf-saving logic of `huggingface_hub` (i.e. the existing `export_folder_to_dduf`)

And we document that loading happens through:

```py
pipeline = FluxPipeline.from_pretrained(<model_id>, dduf_file="filename.dduf")
```

and saving happens with

```py
pipeline.save_pretrained("path/to/local/folder")
export_folder_to_dduf("path/to/local/folder", "my_dduf_file.dduf")
```

This way people can play with it and we don't constrain ourselves for future development on how we want to implement saving in `diffusers`.

sayakpaul commented:
Yessss! Perfectly fine with this approach. Thank you for explaining.

Wauplin commented:
@julien-c @LysandreJik @sayakpaul @SunMarc @hanouticelina I just completed the tests and fixed some bugs and docstrings. I also added some basic validation rules to the parser (e.g. no nested directories). I think this PR is now ready for review before been merged.

Can you all get a review and confirm that the current interface (`read_dduf_file` / `export_*`, `DDUFEntry`) and current file format is ok for you? Once this is merged, I'll open a follow-up PR to document all this :)

hanouticelina commented on src/huggingface_hub/serialization/_dduf.py:
```diff
@@ -0,0 +1,354 @@
+    """
+    entries = {}
+    dduf_path = Path(dduf_path)
+    logger.info("Reading DDUF file %s", dduf_path)
```
(nit) I'm leaning more towards f-strings for readability and consistency with the rest of the codebase
```suggestion
    logger.info(f"Reading DDUF file {dduf_path}")
```

Wauplin replied:
Old habit of mine :smile:. This [comment](https://www.reddit.com/r/learnpython/comments/xx14jw/comment/ir9gjex/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button) explains quite well why `%s` is considered better than f-string for logs. Agree though that this is a very micro-optimization and that consistency with the existing codebase is more important.

> _It is possible to log a message, but for that message never to actually appear in any logs - for example, because it is a lower level than the logger is set to (eg DEBUG when the logger is at INFO). So there is no point doing the work to render all the contents of that message - which might potentially involve calling some expensive methods - if it is never going to be output._

Wauplin replied:
addressed in https://github.com/huggingface/huggingface_hub/pull/2692/commits/dca1586c901de2c44de50c562c21bbd5ee2f81b3

hanouticelina commented on src/huggingface_hub/serialization/_dduf.py:
```diff
@@ -0,0 +1,354 @@
+            entries[info.filename] = DDUFEntry(
+                filename=info.filename, offset=offset, length=info.file_size, dduf_path=dduf_path
+            )
+    logger.info("Done reading DDUF file %s. Found %d entries", dduf_path, len(entries))
```
feel free to keep %-formatting if there's a specific reason for it
```suggestion
    logger.info(f"Done reading DDUF file {dduf_path}. Found {len(entries)} entries")
```

hanouticelina replied:
same comment for the other logging statements

hanouticelina commented on src/huggingface_hub/serialization/_dduf.py:
````diff
@@ -0,0 +1,354 @@
+    Example:
+        ```python
+        >>> import json
+        >>> import safetensors.load
````
```suggestion
        >>> import safetensors
```

Wauplin replied:
```suggestion
        >>> import safetensors.torch
```

hanouticelina commented on src/huggingface_hub/serialization/_dduf.py:
````diff
@@ -0,0 +1,354 @@
+    Example:
+        ```python
+        >>> from huggingface_hub import export_folder_as_dduf
+        >>> export_folder_as_dduf("FLUX.1-dev.dduf", diffuser_path="path/to/FLUX.1-dev")
````
```suggestion
        >>> export_folder_as_dduf("FLUX.1-dev.dduf", folder_path="path/to/FLUX.1-dev")
```

hanouticelina commented on src/huggingface_hub/serialization/_dduf.py:
```diff
@@ -0,0 +1,354 @@
+
+    def _iterate_over_folder() -> Iterable[Tuple[str, Path]]:
+        for path in Path(folder_path).glob("**/*"):
+            if path.is_dir():
```
(nit) feel free to ignore, it's just for clearer intent
```suggestion
            if not path.is_file():
```

hanouticelina commented on src/huggingface_hub/serialization/_dduf.py:
````diff
@@ -0,0 +1,354 @@
+        ```
+    """
+    logger.info("Exporting DDUF file '%s'", dduf_path)
+    filenames = set()
````
👍  using a set() for tracking the files is much better than calling `archive.namelist()` each time! according to the documentation, it reparses each file header in the directory for each lookup.

hanouticelina reviewed (commented):
looks good and clean to me, thanks @Wauplin ! left some comments, nothing major

pcuenca commented on docs/source/en/package_reference/serialization.md:
```diff
@@ -6,6 +6,116 @@ rendered properly in your Markdown viewer.
+...     # Build an generator that yields the entries to add to the DDUF file.
+...     # The first element of the tuple is the filename in the DDUF archive (must use UNIX separator!). The second element is the content of the file.
+...     # Entries will be evaluated lazily when the DDUF file is created (only 1 entry is loaded in memory at a time)
+...     yield "vae/config.json", pipe.vae.to_json_string().encode()
```
I was confused that this is `pipe.vae.to_json_string()` but then it's `pipe.text_encoder.config.to_json_string()` below, but I think it's correct as the vae config is a FrozenDict.

Wauplin replied:
I'd be up for another solution tbh. I just wanted to show that you need to serialize things as bytes by yourself.

sayakpaul commented on docs/source/en/package_reference/serialization.md:
```diff
@@ -6,6 +6,116 @@ rendered properly in your Markdown viewer.
+# Export specific files from the local disk.
+>>> from huggingface_hub import export_entries_as_dduf
+>>> export_entries_as_dduf(
+...     "stable-diffusion-v1-4-FP16.dduf",
```
Expose the function argument name here, too?

sayakpaul commented on docs/source/en/package_reference/serialization.md:
````diff
@@ -6,6 +6,116 @@ rendered properly in your Markdown viewer.
+>>> export_entries_as_dduf("stable-diffusion-v1-4.dduf", entries=as_entries(pipe))
+```
+
+**Note:** in practice, `diffusers` provides a method to directly serialize a pipeline in a DDUF file. The snippet above is only meant as an example.
````
```suggestion
**Note:** In practice, `diffusers` provides a method to directly serialize a pipeline in a DDUF file. The snippet above is only meant as an example.
```

But we are not providing this utility for saving as of yet no? So, maybe remove?

Wauplin replied:
removed it

sayakpaul commented on docs/source/en/package_reference/serialization.md:
```diff
@@ -6,6 +6,116 @@ rendered properly in your Markdown viewer.
+>>> from huggingface_hub import read_dduf_file
+
+# Read DDUF metadata
+>>> dduf_entries = read_dduf_file("FLUX.1-dev.dduf")
```
Consider a fully working example? Here, I don't know where `Flux.1-dev.dduf` is coming from.

Wauplin replied:
Since the dduf file can come from anywhere (a local export, from `hf_hub_download`, a download from the browser, etc.) it's hard to provide a full example

sayakpaul commented on docs/source/en/package_reference/serialization.md:
```diff
@@ -6,6 +6,116 @@ rendered properly in your Markdown viewer.
+
+# Returns a mapping filename <> DDUFEntry
+>>> dduf_entries["model_index.json"]
+DDUFEntry(filename='model_index.json', offset=66, length=587)
```
Elaborate a bit about `offset` and `length`? Otherwise, I am completely off with what they mean as a user.

Wauplin replied:
I could remove them from the example but I added them because that's what printed in the terminal. I prefer not to document further because since it's not really important compared to `read_text` and `as_mmap` methods shows below. The attributes are described in the `DDUFEntry` class docstring

sayakpaul commented on docs/source/en/package_reference/serialization.md:
```diff
@@ -6,6 +6,116 @@ rendered properly in your Markdown viewer.
+{'_class_name': 'FluxPipeline', '_diffusers_version': '0.32.0.dev0', '_name_or_path': 'black-forest-labs/FLUX.1-dev', 'scheduler': ['diffusers', 'FlowMatchEulerDiscreteScheduler'], 'text_encoder': ['transformers', 'CLIPTextModel'], 'text_encoder_2': ['transformers', 'T5EncoderModel'], 'tokenizer': ['transformers', 'CLIPTokenizer'], 'tokenizer_2': ['transformers', 'T5TokenizerFast'], 'transformer' ...
+
+# Load VAE weights using safetensors
+>>> with dduf_entries["vae/diffusion_pytorch_model.safetensors"].as_mmap() as mm:
```
```suggestion
>>> import safetensors.torch
>>> with dduf_entries["vae/diffusion_pytorch_model.safetensors"].as_mmap() as mm:
```

Wauplin replied:
(already included above)

sayakpaul commented on docs/source/en/package_reference/serialization.md:
```diff
@@ -6,6 +6,116 @@ rendered properly in your Markdown viewer.
 
+## DDUF file format
+
+DDUF is a file format designed for diffusion models. It allows saving all the information to run a model in a single file. This work is inspired by the GGUF format. `huggingface_hub` provides helpers to save and load DDUF files, ensuring the file format is respected.
```
Do we wanna also talk a bit about the validation approach we have taken and that if need be we will update it?

sayakpaul commented on tests/test_dduf.py:
```diff
@@ -0,0 +1,180 @@
+
+        (folder_path / "config.json").touch()
+        (folder_path / "model.safetensors").touch()
+        (folder_path / "model.bin").touch()  # won't be included
```
If these are not included, we're logging that info, right? In that case, should we also test if the info is logged? Sorry if it's already there.

Wauplin replied:
There are some `DEBUG` logs yes but not worth testing them IMO

sayakpaul replied:
Will agree to disagree here but really up to you. No strong onions.

sayakpaul reviewed (approved):
Thanks, Lucain! My comments are mostly minor in nature. I think this is ready to 🚢

hanouticelina commented on docs/source/en/package_reference/serialization.md:
````diff
@@ -6,6 +6,116 @@ rendered properly in your Markdown viewer.
+
+```python
+>>> import json
+>>> import safetensors.load
````
```suggestion
>>> import safetensors.torch
```
(related to Sayak's comment below)

SunMarc commented on src/huggingface_hub/serialization/_dduf.py:
```diff
@@ -0,0 +1,387 @@
+    "config.json",
+    "tokenizer_config.json",
+    "image_processor.json",
+    "scheduler_config.json"
```
added this one also

SunMarc commented on src/huggingface_hub/serialization/_dduf.py:
```diff
@@ -0,0 +1,387 @@
+                logger.debug(f"Skipping file '{path}' (file type not allowed)")
+                continue
+            path_in_archive = path.relative_to(folder_path)
+            if len(path_in_archive.parts) >= 3:
+                logger.debug(f"Skipping file '{path}' (nested directories not allowed)")
+                continue
```
should we raise an error instead here ?

Wauplin replied:
I'd rather not. For instance if there's a README in a repo, raising an error is a bit too extreme IMO

Wauplin replied:
(but why setting the log level higher in the future if we see it's a recurrent complain)

SunMarc reviewed (approved):
LGTM ! Just a nit

sayakpaul commented on src/huggingface_hub/serialization/_dduf.py:
```diff
@@ -0,0 +1,387 @@
+    # Each folder must contain at least one of these entries
+    "config.json",
+    "tokenizer_config.json",
+    "image_processor.json",
```
```suggestion
    "preprocessor_config.json",
```

Sincere apologies.

Wauplin replied:
no worries, updated in https://github.com/huggingface/huggingface_hub/pull/2692/commits/6ad29e77104d28d07673459678fef38fbaad34bf

Wauplin commented:
Thanks everyone for the feedback!
